pkg/storegateway/indexheader/index/postings.go

```diff
@@ -448,36 +503,67 @@ func (t *PostingOffsetTableV2) LabelValues(name string, prefix string, filter fu
 			d.Skip(skip)
 		}
 
-		unsafeValue := yoloString(d.UnsafeUvarintBytes())
-		if prefix == "" {
-			// Quick path for no prefix matching.
-			if filter == nil || filter(unsafeValue) {
-				// Clone the yolo string since its bytes will be invalidated as soon as
-				// any other reads against the decoding buffer are performed.
-				values = append(values, strings.Clone(unsafeValue))
-			}
+		val = yoloString(d.UnsafeUvarintBytes())
+
+		prefixMatches := strings.HasPrefix(val, prefix)
+		isAMatch = prefixMatches && (filter == nil || filter(val))
+		noMoreMatches = val == lastVal || (!prefixMatches && prefix < val)
+		// Clone the yolo string since its bytes will be invalidated as soon as
+		// any other reads against the decoding buffer are performed.
+		// We'll only need the sting if it matches our filter.
+		if isAMatch {
+			val = strings.Clone(val)
 		} else {
-			if strings.HasPrefix(unsafeValue, prefix) {
-				if filter == nil || filter(unsafeValue) {
-					// Clone the yolo string since its bytes will be invalidated as soon as
-					// any other reads against the decoding buffer are performed.
-					values = append(values, strings.Clone(unsafeValue))
-				}
-			} else if prefix < unsafeValue {
-				// There will be no more values with the prefix.
-				break
-			}
+			val = ""
+		}
+		// The information in length and number of entries is redundant,
+		// so we can omit the first one - length - and return
+		// the offset of the number_of_entries field.
+		startOff = int64(d.Uvarint64()) + postingLengthFieldSize
+		return
+	}
+
+	var (
+		currList         PostingListOffset
+		nextIsConsumed   bool
+		nextValueSafe    string
+		nextValueMatches bool
+		nextOffset       int64
+		noMoreMatches    bool
+	)
+
+	for d.Err() == nil {
+		currentValueIsLast := noMoreMatches
+		currentValueMatches := nextValueMatches
+		if nextIsConsumed {
```

Comment:
[nit] A little comment to make it easier to understand:

```suggestion
		// Populate the current list either reading it from the pre-populated "next" or reading it from index.
		if nextIsConsumed {
```